network/endpoint_linux.go

```diff
@@ -54,204 +55,203 @@ func (nw *network) newEndpointImpl(
+						if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
+							nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
+						}
+						epClient = NewTransparentVlanEndpointClient(nw, epInfo, hostIfName, contIfName, vlanid, localIP, nl, plc)
+					} else {
+						log.Printf("OVS client")
+						if _, ok := epInfo.Data[SnatBridgeIPKey]; ok {
+							nw.SnatBridgeIP = epInfo.Data[SnatBridgeIPKey].(string)
+						}
+
+						epClient = NewOVSEndpointClient(
+							nw,
+							epInfo,
+							hostIfName,
+							contIfName,
+							vlanid,
+							localIP,
+							nl,
+							ovsctl.NewOvsctl(),
+							plc)
+					}
+				} else if nw.Mode != opModeTransparent {
+					log.Printf("Bridge client")
+					epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
+				} else {
+					log.Printf("Transparent client")
+					epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
+				}
 			}
-		} else if nw.Mode != opModeTransparent {
-			log.Printf("Bridge client")
-			epClient = NewLinuxBridgeEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
-		} else {
-			log.Printf("Transparent client")
-			epClient = NewTransparentEndpointClient(nw.extIf, hostIfName, contIfName, nw.Mode, nl, plc)
 		}
-	}
 
-	// Cleanup on failure.
-	defer func() {
-		if err != nil {
-			log.Printf("CNI error:%v. Delete Endpoint %v and rules that are created.", err, contIfName)
-			endpt := &endpoint{
-				Id:                       epInfo.Id,
-				IfName:                   contIfName,
-				HostIfName:               hostIfName,
-				LocalIP:                  localIP,
-				IPAddresses:              epInfo.IPAddresses,
-				DNS:                      epInfo.DNS,
-				VlanID:                   vlanid,
-				EnableSnatOnHost:         epInfo.EnableSnatOnHost,
-				EnableMultitenancy:       epInfo.EnableMultiTenancy,
-				AllowInboundFromHostToNC: epInfo.AllowInboundFromHostToNC,
-				AllowInboundFromNCToHost: epInfo.AllowInboundFromNCToHost,
+		//nolint:gocritic
+		defer func(client EndpointClient, contIfName string) {
+			// Cleanup on failure.
+			if err != nil {
+				log.Printf("CNI error:%v. Delete Endpoint %v and rules that are created.", err, contIfName)
+				if containerIf != nil {
+					client.DeleteEndpointRules(ep)
+				}
+				// set deleteHostVeth to true to cleanup host veth interface if created
+				//nolint:errcheck // ignore error
+				client.DeleteEndpoints(ep)
 			}
+		}(epClient, contIfName)
 
-			if containerIf != nil {
-				endpt.MacAddress = containerIf.HardwareAddr
-				epClient.DeleteEndpointRules(endpt)
+		err = func() error {
+			if epErr := epClient.AddEndpoints(epInfo); epErr != nil {
+				return epErr
 			}
 
-			if nw.extIf != nil {
-				endpt.Gateways = []net.IP{nw.extIf.IPv4Gateway}
+			if epInfo.AddressType == cns.Default {
+				var epErr error
+				containerIf, epErr = netioCli.GetNetworkInterfaceByName(contIfName)
+				if epErr != nil {
+					return epErr
```

Comment:
this check should be moved inside AddEndpoints and lets not add anything specific on this (comparing default or secondary and then take actions)

Reply:
lets keep this part here for now to minimize changes since we'd have to update all the other clients to modify endpoint struct

Reply:
we still have to set the endpoint MacAddress based on the results

ep.MacAddress = containerIf.HardwareAddr

and the defer also checks containerIf before calling client.DeleteEndpointRules

Reply:
I think we can get container mac from the ep client but still need to check for default address type

ep.MacAddress = epClient.containerMac

In this case can we check if mac is empty to call epClient.DeleteEndpointRules?

Reply:
lets discuss on this offline